plugins/woocommerce/src/Internal/Traits/RestApiCache.php

```diff
@@ -0,0 +1,405 @@
+ * - Cached response TTL expires.
+ *
+ * NOTE: This caching mechanism uses the WordPress cache (wp_cache_* functions).
+ * By default caching is only enabled when an external object cache is enabled
+ * (checked via call to VersionStringGenerator::can_use()), so the cache is persistent
+ * across requests and not just for the current request.
+ *
+ * @since   10.4.0
+ */
+trait RestApiCache {
+	/**
+	 * Cache group name for REST API responses.
+	 *
+	 * @var string
+	 */
+	private static string $cache_group = 'woocommerce_rest_api_cache';
+
+	/**
+	 * Cache TTL in seconds.
+	 *
+	 * @var int
+	 */
+	private static int $cache_ttl = HOUR_IN_SECONDS;
+
+	/**
+	 * The instance of VersionStringGenerator to use, or null if caching is disabled.
+	 *
+	 * @var VersionStringGenerator|null
+	 */
+	private ?VersionStringGenerator $version_string_generator = null;
+
+	/**
+	 * Initialize the trait.
+	 * This MUST be called from the controller's constructor.
+	 */
+	protected function initialize_rest_api_cache(): void {
+		$generator                      = wc_get_container()->get( VersionStringGenerator::class );
+		$this->version_string_generator = $generator->can_use() ? $generator : null;
+	}
+
+	/**
+	 * Wrap an endpoint callback declaration with caching logic.
+	 * Usage: `'callback' => $this->with_cache( array( $this, 'endpoint_callback_method' ) )`
+	 *        `'callback' => $this->with_cache( array( $this, 'endpoint_callback_method' ), [ 'entity_type' => 'product' ] )`
+	 *
+	 * @param callable $callback The original endpoint callback.
+	 * @param array    $config   Caching configuration:
+	 *                           - entity_type: string (falls back to get_default_entity_type()).
+	 * @return callable Wrapped callback.
+	 */
+	protected function with_cache( callable $callback, array $config = array() ): callable {
+		return fn( $request ) => $this->handle_cacheable_request( $request, $callback, $config );
+	}
+
+	/**
+	 * Handle a request with caching logic.
+	 *
+	 * Strategy: Try to use cached response if available and valid, otherwise execute the endpoint
+	 * callback and cache the response (if successful) for future requests.
+	 *
+	 * @param WP_REST_Request $request  The request object.
+	 * @param callable        $callback The original endpoint callback.
+	 * @param array           $config   Caching configuration specified for the endpoint.
+	 * @return WP_REST_Response|\WP_Error The response.
+	 */
+	private function handle_cacheable_request( WP_REST_Request $request, callable $callback, array $config ) {
+		if ( is_null( $this->version_string_generator ) ) {
+			return call_user_func( $callback, $request );
+		}
+
+		$should_skip_cache = ! $this->should_use_cache_for_request( $request );
+		if ( ! $should_skip_cache ) {
+			$cached_config     = $this->build_cache_config( $request, $config );
+			$should_skip_cache = is_null( $cached_config );
+		}
+
+		if ( $should_skip_cache ) {
```

Comment:
Nitpick, but we could simplify this into an `if/else`?

Reply:
Noting that this can't be refactored to an if-else as suggested, because `$should_skip_cache` might be modified insite the  first ìf` block.